### PR TITLE
Add local search

### LINKED ISSUE #22: Enable search
Currently we don't have search functionality in the docs site, which would be very useful.

[Docusaurus supports it with various methods](https://docusaurus.io/docs/search), so let's look into what would work best for us.

We don't have a lot of documentation, so hopefully a local on-device index would be feasible. Otherwise let's consider one of the alternatives.

### PR DESCRIPTION
**Pull request by merlinfuchs**
This adds local search using https://github.com/easyops-cn/docusaurus-search-local for the docs and blog. The search works across v1 and v2 which can be confusing sometimes, but I haven't found a way to scope it to the currently selected version.

There are a few options to configure this, most notably if we want to include the blog articles in the search index or not.

Bildschirmfoto 2024-10-22 um 10 17 41
Bildschirmfoto 2024-10-22 um 10 17 52

closes #22

### CODE REVIEW

File: docusaurus.config.js [hunk 1]
```diff
@@ -38,6 +38,17 @@ const config = {
 
   plugins: ["docusaurus-plugin-sass"],
 
+  themes: [
+    [
+      require.resolve("@easyops-cn/docusaurus-search-local"),
+      /** @type {import("@easyops-cn/docusaurus-search-local").PluginOptions} */
+      {
+        hashed: true,
+        indexPages: true,
```
Comment (merlinfuchs) [inline docusaurus.config.js:47]: I have added this so the integrations page, which isn't part of the docs, is also indexed.

File: package-lock.json [hunk 2]
```diff
@@ -17300,6 +17704,135 @@
...
+        "tslib": "^2.4.0"
+      }
+    },
+    "@emnapi/wasi-threads": {
+      "version": "1.0.1",
+      "resolved": "https://registry.npmjs.org/@emnapi/wasi-threads/-/wasi-threads-1.0.1.tgz",
+      "integrity": "sha512-iIBu7mwkq4UQGeMEM8bLwNK962nXdhodeScX4slfQnRhEMMzvYivHhutCIk8uojvmASXXPC2WNEjwxFWk72Oqw==",
+      "optional": true,
+      "requires": {
+        "tslib": "^2.4.0"
```
Comment (imiric) [inline package-lock.json:17833]: :thinking: Do you know why these `@emnapi`, and `@napi-rs` and `@node-rs` dependencies below are added here?

They're not specified in `easyops-cn/docusaurus-search-local`'s `package.json`, and `npm explain` wasn't helpful. Some of them are not installed in `node_modules`, but others like `@node-rs/jieba` are. That package pulls in some binaries related to https://github.com/messense/jieba-rs, which is a word segmentation package for Chinese implemented in Rust.

Ah, these are part of the `yarn.lock` file, which seems to also be used when resolving dependencies. The Node ecosystem is such a security nightmare... :face_exhaling:

It sucks that we have to depend on these packages even though we likely won't ever use Chinese in our docs. In general, I distrust Chinese software, perhaps unreasonably so, especially if it comes in binary blobs like this. Should we consider using [cmfcmf/docusaurus-search-local](https://github.com/cmfcmf/docusaurus-search-local) instead, which the `easyops-cn` package is a fork of? The changes from the fork don't seem that relevant for our use case.

**Review by imiric — APPROVED** [review]
Thanks for hooking this up! :tada:

I tested it locally and it works great. I wondered if we should include the blog in the index or not, but it might be useful to get results from there as well, so let's leave it enabled.

I just have a minor concern about the number of changes in the `package-lock.json` file, but it's not a blocker.

File: package-lock.json (hunk 2 above)
Comment (merlinfuchs) [reply package-lock.json:17833]: The reasons I chose `easyops-cn/docusaurus-search-local` over `cmfcmf/docusaurus-search-local` were mostly cosmetic. I think it looks a lot nicer but I'm also happy to try out `cmfcmf/docusaurus-search-local` and maybe tweak the design a bit 🤔

**Review by greenberga — APPROVED** [review]
Nice!

File: package-lock.json (hunk 2 above)
Comment (imiric) [reply package-lock.json:17833]: Yeah, the `easyops-cn` package does look nicer. If the difference with the `cmfcmf` one is substantial, then let's not spend time tweaking the design. In all likelihood there's no reason to worry about the `easyops-cn` package, so let's keep it. It's just that seeing these binary deps does tingle my spidey senses :sweat_smile:, even if they're not part of the final bundle...

**Pull request merged**